arwen/contexts/async.go

```diff
@@ -516,65 +574,104 @@ func (context *asyncContext) addAsyncCall(groupID string, call *arwen.AsyncCall)
 	}
 
-	logAsync.Trace("async.Execute() complete locals")
+	return nil
+}
 
-	// This call to closeCompletedAsyncCall() is necessary to remove the
-	// AsyncCall that has been just before async.Execute() was called, within
-	// host.callSCMethod(). This happens when a cross-shard callback returns and
-	// finalizes an AsyncCall.
-	context.closeCompletedAsyncCalls()
-	if context.groupCallbacksEnabled {
-		context.executeCompletedGroupCallbacks()
+// UpdateCurrentAsyncCallStatus detects the AsyncCall returning as callback,
+// extracts the ReturnCode from data provided by the destination call, and updates
+// the status of the AsyncCall with its value.
+func (context *asyncContext) UpdateCurrentAsyncCallStatus(
+	address []byte,
+	callID []byte,
+	vmInput *vmcommon.VMInput,
+) (*arwen.AsyncCall, error) {
+	if vmInput.CallType != vm.AsynchronousCallBack {
+		return nil, nil
 	}
-	context.deleteCompletedGroups()
 
-	logAsync.Trace("async.Execute() execute remote")
-	// Step 2: in one combined step, do the following:
-	// * locally execute built-in functions with cross-shard
-	//   destinations, whereby the cross-shard OutputAccount entries are generated
-	// * call host.sendAsyncCallCrossShard() for each pending AsyncCall, to
-	//   generate the corresponding cross-shard OutputAccount entries
-	// Note that all async calls below this point are pending by definition.
-	for _, group := range context.asyncCallGroups {
-		for _, call := range group.AsyncCalls {
-			err = context.executeAsyncCall(call)
-			if err != nil {
-				return err
-			}
-		}
+	if len(vmInput.Arguments) == 0 {
+		return nil, arwen.ErrCannotInterpretCallbackArgs
 	}
 
-	context.deleteCallGroupByID(arwen.LegacyAsyncCallGroupID)
-	if !context.HasPendingCallGroups() && context.contextCallbackEnabled {
-		logAsync.Trace("async.Execute() execute context callback")
-		context.executeContextCallback()
+	loadedContext, err := readAsyncContextFromStorage(
+		context.host.Storage(),
+		address,
+		context.callbackAsyncInitiatorCallID)
+	if err != nil {
+		return nil, err
 	}
 
-	logAsync.Trace("async.Execute() save")
-	err = context.Save()
+	call, _, _, err := loadedContext.GetAsyncCallByCallID(callID)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	return nil
+	// The first argument of the callback is the return code of the destination call
+	destReturnCode := big.NewInt(0).SetBytes(vmInput.Arguments[0]).Uint64()
+	call.UpdateStatus(vmcommon.ReturnCode(destReturnCode))
+
+	return call, nil
+}
+
+func (context *asyncContext) isMultiLevelAsync(call *arwen.AsyncCall) bool {
+	// ESDTTransferOnCallback must be allowed as an exception, even if it appears
+	// to be a 2-level async call.
+	return context.isCallAsyncOnStack() && call.ExecutionMode != arwen.ESDTTransferOnCallBack
+}
+
+func (context *asyncContext) isCallAsyncOnStack() bool {
+	if context.isCallAsync() {
+		return true
+	}
+
+	for index := range context.stateStack {
+		index = len(context.stateStack) - 1 - index
```

Comment:
what ? is this really needed? for index := range and you modify index again ? does not seem safe.

Reply:
Iterate over an array in reverse

Reply:
Please use:
```
for index := len(context.stateStack) - 1; index >=0; index-- {
    ...
}
```

Reply:
Replaced.